File: src/main/java/com/dafttech/terra/game/world/items/ItemRainbowGun.java
package com.dafttech.terra.game.world.items;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.game.world.entities.EntityRainbow;
import com.dafttech.terra.game.world.entities.models.EntityLiving;
import com.dafttech.terra.resources.Resources;

public class ItemRainbowGun extends ItemEntitySpawner {
    @Override
    public boolean spawnEntity(EntityLiving causer, Vector2 position) {
        EntityRainbow a = new EntityRainbow(causer.getPosition(), causer.worldObj);
        a.setVelocity(Vector2.getMouse().sub(new Vector2(Gdx.graphics.getWidth() / 2, Gdx.graphics.getHeight() / 2)).mul(0.08f));
        return true;
    }

    @Override
    public int getUsedItemNum(EntityLiving causer, Vector2 position) {
        return 0;
    }

    @Override
    public float getNextUseDelay(EntityLiving causer, Vector2 position, boolean leftClick) {
        return 0;
    }

    @Override
    public TextureRegion getImage() {
        return Resources.ITEMS.getImage("rainbowgun");
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/items/ItemRainbowGun.scala
package com.dafttech.terra.game.world.items

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.input.MousePos
import com.dafttech.terra.engine.vector.Vector2d
import com.dafttech.terra.game.world.entities.EntityRainbow
import com.dafttech.terra.game.world.entities.models.EntityLiving
import com.dafttech.terra.resources.Resources
import monix.eval.Task

class ItemRainbowGun extends ItemEntitySpawner {
  override def spawnEntity(causer: EntityLiving, position: Vector2d): Boolean = {
    val a = new EntityRainbow(causer.getPosition)(causer.world)
    a.setVelocity((MousePos.vector2d - (Gdx.graphics.getWidth / 2, Gdx.graphics.getHeight / 2)) * 0.08)
    val time = (System.currentTimeMillis() % 1000) / 100f
    val offset120 = Math.PI * 2 / 3
    val r = Math.max(Math.sin(time + offset120 * 0).toFloat, 0)
    val g = Math.max(Math.sin(time + offset120 * 1).toFloat, 0)
    val b = Math.max(Math.sin(time + offset120 * 2).toFloat, 0)
    a.lightColor = new Color(r, g, b, 0.5f)
    true
  }

  override def getUsedItemNum(causer: EntityLiving, position: Vector2d): Int = 0

  override def getNextUseDelay(causer: EntityLiving, position: Vector2d, leftClick: Boolean): Double = 0

  override val getImage: Task[TextureRegion] = Resources.ITEMS.getImageTask("rainbowgun")
}